src/txreconciliation.cpp

```diff
@@ -0,0 +1,1005 @@
+     * conservation and avoid overflows, we should reconcile in the same order, because then it’s
+     * easier to estimate set difference size.
+     */
+    std::deque<NodeId> m_queue GUARDED_BY(m_mutex);
+
+    /**
+     * Reconciliations are requested periodically:
+     * every RECON_REQUEST_INTERVAL seconds we pick a peer from the queue.
+     */
+    std::chrono::microseconds m_next_recon_request GUARDED_BY(m_mutex);
+    void UpdateNextReconRequest(std::chrono::microseconds now) EXCLUSIVE_LOCKS_REQUIRED(m_mutex)
+    {
+        assert(m_queue.size() > 0);
+        m_next_recon_request = now + RECON_REQUEST_INTERVAL / m_queue.size();
+    }
+
+    /**
+     * Used to schedule the next initial response for any pending reconciliation request.
+     * Respond to all requests at the same time to prevent transaction possession leak.
+     */
+    std::chrono::microseconds m_next_recon_respond{0};
+    std::chrono::microseconds NextReconRespond()
+    {
+        auto current_time = GetTime<std::chrono::microseconds>();
+        if (m_next_recon_respond < current_time) {
+            m_next_recon_respond = current_time + RECON_RESPONSE_INTERVAL;
+        }
+        return m_next_recon_respond;
+    }
+
+    void HandleInitialSketch(std::unordered_map<NodeId, ReconciliationState>::iterator& recon_state,
+        const std::vector<uint8_t>& skdata,
+        // returning values
+        std::vector<uint32_t>& txs_to_request, std::vector<uint256>& txs_to_announce, std::optional<bool>& result)
+    {
+        assert(recon_state->second.m_we_initiate);
+        assert(recon_state->second.m_state_init_by_us.m_phase == Phase::INIT_REQUESTED);
+
+        uint16_t remote_sketch_capacity = uint16_t(skdata.size() / BYTES_PER_SKETCH_CAPACITY);
+        Minisketch local_sketch = recon_state->second.ComputeBaseSketch(remote_sketch_capacity);
+        Minisketch remote_sketch;
+        if (remote_sketch_capacity != 0) {
+            remote_sketch = Minisketch(RECON_FIELD_SIZE, 0, remote_sketch_capacity).Deserialize(skdata);
+        }
+
+        // Remote sketch is empty in two cases per which reconciliation is pointless:
+        // 1. the peer has no transactions for us
+        // 2. we told the peer we have no transactions for them while initiating reconciliation.
+        // In case (2), local sketch is also empty.
+        if (remote_sketch_capacity == 0 || !remote_sketch || !local_sketch) {
+
+            // Announce all transactions we have.
+            txs_to_announce = recon_state->second.m_local_set.GetAllTransactions();
+
+            // Update local reconciliation state for the peer.
+            recon_state->second.FinalizeInitByUs(true, DEFAULT_RECON_Q);
+            recon_state->second.m_state_init_by_us.m_phase = Phase::NONE;
+
+            result = false;
+
+            LogPrint(BCLog::NET, "Reconciliation we initiated with peer=%d terminated due to empty sketch." /* Continue */
```

Comment:
linter not happy with this line:

`All calls to LogPrintf() and LogPrint() should be terminated with \n`